spring-integration-gcp/src/main/java/org/springframework/integration/gcp/storage/GCSSession.java

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+
+	@Override
+	public void write(InputStream inputStream, String destination) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(destination);
+		BlobInfo gcsBlobInfo = BlobInfo.newBuilder(BlobId.of(tokens[0], tokens[1])).build();
+
+		try (InputStream is = inputStream) {
+			try (WriteChannel channel = this.gcs.writer(gcsBlobInfo)) {
+				channel.write(ByteBuffer.wrap(StreamUtils.copyToByteArray(is)));
+			}
+		}
+	}
+
+	@Override
+	public void append(InputStream inputStream, String destination) throws IOException {
+		// TODO(joaomartins): We could do compose here, but it assumes that InputStream is first copied to a
+		// GCS object that we can then "compose" to the original object.
+		throw new UnsupportedOperationException("Appending isn't supported by Google Cloud Storage.");
+	}
+
+	@Override
+	public boolean mkdir(String directory) throws IOException {
+		try {
+			this.gcs.create(BucketInfo.of(directory));
+			return true;
+		}
+		catch (StorageException se) {
+			LOGGER.info("Error creating the GCS bucket.", se);
+			return false;
+		}
+
+	}
+
+	@Override
+	public boolean rmdir(String directory) throws IOException {
+		return this.gcs.delete(directory);
+	}
+
+	@Override
+	public void rename(String pathFrom, String pathTo) throws IOException {
+		String[] fromTokens = getBucketAndObjectFromPath(pathFrom);
+		String[] toTokens = getBucketAndObjectFromPath(pathTo);
+
+		// There is currently no way to rename/move things in GCS, so we'll have to copy and remove.
+		BlobId source = BlobId.of(fromTokens[0], fromTokens[1]);
+		BlobId target = BlobId.of(toTokens[0], toTokens[1]);
+		Storage.CopyRequest copyRequest = Storage.CopyRequest.of(source, target);
+		this.gcs.copy(copyRequest);
+
+		this.gcs.delete(source);
+	}
+
+	@Override
+	public void close() {
+		// GCS session doesn't need to be closed.
+	}
+
+	@Override
+	public boolean isOpen() {
+		return true;
+	}
+
+	@Override
+	public boolean exists(String path) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(path);
+
+		if (tokens.length == 1) {
+			return this.gcs.get(path) != null;
+		}
+
+		return this.gcs.get(tokens[0], tokens[1]) != null;
+	}
+
+	@Override
+	public String[] listNames(String path) throws IOException {
+		List<String> names = Stream.of(list(path))
+				.map(BlobInfo::getName)
+				.collect(Collectors.toList());
+
+		return names.toArray(new String[names.size()]);
+	}
+
+	@Override
+	public InputStream readRaw(String source) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(source);
+		return Channels.newInputStream(this.gcs.reader(tokens[0], tokens[1]));
+	}
+
+	@Override
+	public boolean finalizeRaw() throws IOException {
+		return true;
+	}
+
+	@Override
+	public Object getClientInstance() {
+		return this.gcs;
+	}
+
+	private String[] getBucketAndObjectFromPath(String path) throws IOException {
+		// Assumes paths of the form /bucket/folder/blob
+		String[] tokens = path.split("/", 2);
```

Comment:
It would be good to have a way of setting the separator here, since `AbstractRemoteFileStreamingMessageSource` allows for different separators, but its conceptually disconnected from our `GcsSession`..